NotificationServiceExtension/MultisigNotification+ContentProvider.swift

```diff
@@ -0,0 +1,197 @@
+//  MultisigNotification+ContentProvider.swift
+//  Multisig
+//
+//  Created by Dmitry Bespalov on 26.07.21.
+//  Copyright © 2021 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import CoreData
+import SwiftCryptoTokenFormatter
+
+protocol NotificationContentProvider {
+    func loadContent(completion: @escaping ((title: String, body: String)?) -> Void)
+}
+
+extension MultisigNotification: NotificationContentProvider {
+    func loadContent(completion: @escaping ((title: String, body: String)?) -> Void) {
+        switch self {
+        case let .incomingNativeCoin(n):
+            n.loadContent(completion: completion)
+        case let .incomingToken(n):
+            n.loadContent(completion: completion)
+        case let .executedMultisigTransaction(n):
+            n.loadContent(completion: completion)
+        case let .newConfirmation(n):
+            n.loadContent(completion: completion)
+        case let .confirmationRequest(n):
+            n.loadContent(completion: completion)
+        case .unknown:
+            completion(nil)
+        }
+    }
+}
+
+extension SafeNotification {
+    func loadSafe(completion: @escaping (Safe?, NSManagedObjectContext) -> Void) {
+        NotificationService.coreData.persistentContainer.performBackgroundTask { context in
+            completion(safe(context: context, address: address, chainId: chainId), context)
+        }
+    }
+
+    func safe(context: NSManagedObjectContext, address: AddressString, chainId: UInt256String) -> Safe? {
+        do {
+            let fr = Safe.fetchRequest().by(address: address.description, chainId: chainId.description)
+            let safe = try context.fetch(fr).first
+            return safe
+        } catch {
+            print("Failed to fetch safe: \(error)")
+            return nil
+        }
+    }
+
+    func keyInfo(context: NSManagedObjectContext, address: AddressString) -> KeyInfo? {
+        do {
+            let fr = KeyInfo.fetchRequest().by(address: address.address)
+            let info = try context.fetch(fr).first
+            return info
+        } catch {
+            print("Failed to fetch key info: \(error)")
+            return nil
+        }
+    }
+
+}
+
+extension MultisigNotification.IncomingNativeCoin: NotificationContentProvider {
+    func loadContent(completion: @escaping ((title: String, body: String)?) -> Void) {
+        loadSafe { safeOrNil, _ in
+            guard
+                let safe = safeOrNil,
+                let chain = safe.chain,
+                let chainName = chain.name,
+                let nativeCoin = chain.nativeCurrency,
+                let symbol = nativeCoin.symbol
+            else {
+                completion((
+                    title: "Incoming token (Chain Id \(chainId))",
+                    body: "\(address.address.truncatedInMiddle): \(value) received"
+                ))
+                return
+            }
+            let safeName = safe.name ?? address.address.truncatedInMiddle
```

Comment:
can there be a safe without a name?

Reply:
well, the name is optional, so I handle it here

Reply:
you could use force unwrapping for simplicity